usb: usbip: add initial support for USBIP server

jfischer-no opened the pull request:
The server uses minimal host support to export a USB device to a remote
USBIP client. It supports control and bulk transfers, interrupt
transfers may also work, but this depends on the host controller used.
Only a single USB device can be exported yet.

jfischer-no committed:
usb: usbip: add initial support for USBIP server

The server uses host support to export a USB device to a remote
USBIP client. It supports control and bulk transfers, interrupt
transfers may also work, but this depends on the host controller used.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: implement the dequeue function

Implement the dequeue function and remove the timeout parameter as it is
no longer necessary and can be handled in the upper layer. The dequeue
function is required for the USBIP implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on drivers/usb/uhc/uhc_virtual.c:
```diff
@@ -174,36 +187,89 @@ static int vrt_xfer_bulk(const struct device *dev,
+	unsigned int key;
+	int err;
+
+	for (uint8_t ch = 0; ch < 31; ch++) {
```
Why `< 31`?

jfischer-no replied:
outdated

tmon-nordic commented on subsys/usb/host/usbip.h:
```diff
@@ -0,0 +1,124 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
This file is missing include guards.

jfischer-no committed:
drivers: uhc_virtual: fix SOF timer handling

Do not necessarily call k_timer_init() in the uhc_bus_resume()
implementation. Restart SOF timer after bus reset and in
uhc_sof_enable() implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc_virtual: allow multiple transfers scheduled within a frame

With the current implementation, NACK claims all resources and prevents
scheduling multiple transfers within a frame. Place a number of
available transfers in a list at the beginning of a frame, and process
the transfers in a limited number of slots.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: usb: add initial USB/IP documentation

Add initial USB/IP documentation for the new USB support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc_virtual: update SoF timer based on the bus speed

Update the SoF timer based on the bus speed of the connected device.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: move representation of a USB device to UHC header

This will allow the controller driver to access information about the
device in the future.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: rework usbh_xfer_alloc() parameters

The transfers require enpoint MPS for proper transaction handling,
assign it in the common place during transfer allsocation so that it can
be reused.
Some users, such as USBIP, may need to keep a reference to private data,
add a parameter for completion callback data.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: fix callback parameter type

The parameter should not be a pointer to type usbh_udev_cb_t.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: remove UHC transfer parameter addr and attrib

Address parameter/argument is no longer needed because we have a pointer
to the USB device. The Attrib parameter has never been used and will be
replaced by the interval and start-frame parameters in the future.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: allow partially read device descriptor

To determine bMaxPacketSize0, which is unknown after the reset.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: lock scheduler in usbd_enable()

Lock the scheduler until the device is fully enabled, similar to
usbd_init();

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: interrupt transaction handling

Add interrupt and start frame parameters and schedule transaction at
specific frame. Implement for virtual driver only.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uhc: add opaque pointer to store upper layer private data

Add an opaque pointer to store upper layer private data and initialize
it with the USB host context during controller initialization. Use the
pointer in event processing to get the correct context.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: move bus event handling to a separate thread

Handle them in separate contexts so that the request completion callback
cannot be blocked, such as when a device connection is detected and
requires configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: host: allow dynamic allocation of USB devices

Allow dynamic allocation of USB devices on connected event.
Add very basic USB device validation and configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
snippets: add USB/IP on native simulator snippet

Snippet that helps to build USB device samples with virtual device and
host controllers and USBIP support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/host/usbh_ch9.c:
```diff
@@ -116,10 +116,11 @@ int usbh_req_desc(struct usb_device *const udev,
 {
 	const uint8_t type = USB_DESC_DEVICE;
-	const uint16_t wLength = sizeof(struct usb_device_descriptor);
+	const uint16_t wLength = MIN(len, sizeof(struct usb_device_descriptor));
```
I don't understand this commit at all. If the only place where `usbh_req_desc_dev()` is called, calls this with `sizeof(struct usb_device_descriptor)` then this does not change anything at all.

The whole thing with `bMaxPacketSize0` can be solved by just explicitly reading 8 bytes first. This will always result in single data stage packet regardless of `bMaxPacketSize0`. Then the stack can remember the `bMaxPacketSize0` and handle futher communication without having to guess "was it a short packet, or is bMaxPacketSize0 8".

jfischer-no replied:
After the reset, the device descriptor and bMaxPacketSize0 are unknown. To get it we need to read the first 8 bytes of the descriptor, the 8 is also the minimum value that bMaxPacketSize0 can have. The host controller driver can get bMaxPacketSize0 from the device descriptor. So for the first fetch usbh_device_init() sets both bMaxPacketSize0 and lenght to 8 and calls usbh_req_desc_dev() (which limits both transaction and transfer length to 8 octets).

tmon-nordic replied:
The source of confusion is that the "usb: host: allow partially read device descriptor" doesn't really change anything, but only "usb: host: allow dynamic allocation of USB devices" makes use of the new parameter by calling `usbh_req_desc_dev(udev, 8, &udev->dev_desc);`

raveious commented:
If you don't mind me asking, what environment(s) did you test this on? I'm very interested in this and would like to try and replicate the results.

tmon-nordic commented on drivers/usb/uhc/uhc_common.c:
```diff
@@ -93,20 +93,41 @@ void uhc_xfer_buf_free(const struct device *dev, struct net_buf *const buf)
+			goto xfer_alloc_error;
+		}
+
+		mps = ep_desc->wMaxPacketSize;
```
How is `ep_desc` stored in memory? If it has bytes as-is in the descriptor, then this is missing endian conversion.

jfischer-no replied:
Device descriptor is stored in CPU native endianness.

ddavidebor commented:
I just want to put forward a huge thank you to @jfischer-no for your work on this! I'm really excited and looking forward to this being merged

jfischer-no commented:
rebased to resolve conflicts (loopback PR).

jfischer-no commented:
> I just want to put forward a huge thank you to @jfischer-no for your work on this! I'm really excited and looking forward to this being merged

This is still initial support. There are performance issues using it with virtual devices such as native_sim, and I have not had time to test it on real devices, but that is on my list.

nordicjm commented on subsys/usb/host/Kconfig.usbip:
```diff
@@ -0,0 +1,42 @@
+module-str = usbip
+source "subsys/logging/Kconfig.template.log_config"
+
+endif #USBIP
```
```suggestion
endif # USBIP
```

nordicjm commented on doc/connectivity/usb/host/usbip.rst:
```diff
@@ -0,0 +1,75 @@
+   modprobe usbip-core
+   modprobe usbip-host
+
+There are a few basic commands for everyday use. To list exported USB devices,
```
worth mentioning packages to get for ubuntu here so users don't have to search for why the command is not found?

nordicjm commented on snippets/usbip-native-sim/usbip-native-sim.conf:
```diff
@@ -0,0 +1,22 @@
+CONFIG_NET_TCP=y
+CONFIG_NET_SOCKETS=y
+
+#CONFIG_NO_OPTIMIZATIONS=y
```
```suggestion
```

nordicjm reviewed (dismissed):
Nit. build system changes OK
